### PR TITLE
[clang][ExtractAPI] Compute inherited availability information

### PR DESCRIPTION
**Pull request by daniel-grumberg**
Additionally this computes availability information for all platforms ahead of possibly introducing a flag to enable this behavior.

rdar://123513706

### CODE REVIEW

**Review by QuietMisdreavus — APPROVED** [review]

File: clang/test/ExtractAPI/inherited_availability.m [hunk 1]
```diff
@@ -0,0 +1,149 @@
+// RUN: rm -rf %t
+// RUN: %clang_cc1 -extract-api --pretty-sgf --emit-sgf-symbol-labels-for-testing -triple arm64-apple-macosx \
+// RUN:   -x objective-c-header %s -o %t/output.symbols.json -verify
+
+
+// RUN: FileCheck %s --input-file %t/output.symbols.json --check-prefix A
+__attribute__((availability(macos, introduced=9.0, deprecated=12.0, obsoleted=20.0)))
+@interface A
```
Comment (cyndyishida) [inline clang/test/ExtractAPI/inherited_availability.m:8]: What's the expected availability information emitted when the enclosing entity doesn't have one defined? 
```
@interface A
@property int moreRestrictive __attribute__((availability(macos, introduced=10.0, deprecated=11.0, obsoleted=19.0)));
@end 
```

File: clang/lib/AST/Availability.cpp [hunk 2]
```diff
@@ -16,33 +16,101 @@
...
-  AvailabilityInfo Availability;
+  void insert(clang::AvailabilityInfo &&Availability) {
+    auto *Found = getForPlatform(Availability.Domain);
+    if (Found)
+      Found->mergeWith(std::move(Availability));
+    else
+      Availabilities.emplace_back(std::move(Availability));
+  }
+
+  clang::AvailabilityInfo *getForPlatform(llvm::StringRef Domain) {
```
Comment (cyndyishida) [inline clang/lib/AST/Availability.cpp:34]: nit: why not `std::optional`?

File: clang/lib/AST/Availability.cpp [hunk 3]
```diff
@@ -16,33 +16,101 @@
 #include "clang/AST/Decl.h"
 #include "clang/Basic/TargetInfo.h"
 
-namespace clang {
+namespace {
+
+struct AvailabilitySet {
```
Comment (cyndyishida) [inline clang/lib/AST/Availability.cpp:21]: Can you add a doc string or change the type name to match closer to what this captures? IIUC it's for member attributes' availability

File: clang/lib/AST/Availability.cpp [hunk 4]
```diff
@@ -16,33 +16,101 @@
...
+  if (!Obsoleted.empty() && !Other.Obsoleted.empty())
+    Obsoleted = std::min(Obsoleted, Other.Obsoleted);
+  else
+    Obsoleted = std::max(Obsoleted, Other.Obsoleted);
+}
+
+AvailabilityInfo AvailabilityInfo::createFromDecl(const Decl *D) {
+  AvailabilitySet Availabilities;
+  createInfoForDecl(D, Availabilities);
+  // Traverse
```
Comment (cyndyishida) [inline clang/lib/AST/Availability.cpp:97]: nit: I think this can be removed

**Review by cyndyishida — COMMENTED** [review]
> Additionally this computes availability information for all platforms ahead 

Can you add a test for this?

File: clang/lib/AST/Availability.cpp [hunk 5]
```diff
@@ -16,33 +16,101 @@
 #include "clang/AST/Decl.h"
 #include "clang/Basic/TargetInfo.h"
 
-namespace clang {
+namespace {
+
+struct AvailabilitySet {
+  llvm::SmallVector<clang::AvailabilityInfo> Availabilities;
```
Comment (cyndyishida) [inline clang/lib/AST/Availability.cpp:22]: Can you mark this private, its not clear otherwise that one should be going through `insert`

**Comment by daniel-grumberg** [comment]
> > Additionally this computes availability information for all platforms ahead
> 
> Can you add a test for this?

The behavior is not exposed yet so I can't add a test, will do when I start exposing the behavior.

File: clang/test/ExtractAPI/inherited_availability.m (hunk 1 above)
Comment (daniel-grumberg) [reply clang/test/ExtractAPI/inherited_availability.m:8]: It walks DeclContext's upwards to find one to compute the most restrictive information. If the are no other annotations for that platform walking upwards then it is just the one in the availability annotation. I realized that the existing availability test case doesn't cover this, thanks for pointing it out I will add a test to validate this behavior.

File: clang/lib/AST/Availability.cpp (hunk 2 above)
Comment (daniel-grumberg) [reply clang/lib/AST/Availability.cpp:34]: The semantic is that you get a reference of the thing inside the Availabilities vector if it exist (this method is used by `insert` and not just `AvailabilityInfo::createFromDecl`), to express that with `std::optional<AvailabilityInfo &>` which doesn't exist.

**Review by cyndyishida — APPROVED** [review]
LGTM

**Pull request merged**